File: ipc/ipdl/ipdl/ast.py
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/.

from .util import hash_str


NOT_NESTED = 1
INSIDE_SYNC_NESTED = 2
INSIDE_CPOW_NESTED = 3

NESTED_ATTR_MAP = {
    "not": NOT_NESTED,
    "inside_sync": INSIDE_SYNC_NESTED,
    "inside_cpow": INSIDE_CPOW_NESTED,
}

# Each element of this list is the IPDL source representation of a priority.
priorityList = ["normal", "input", "vsync", "mediumhigh", "control"]

priorityAttrMap = {src: idx for idx, src in enumerate(priorityList)}

NORMAL_PRIORITY = priorityAttrMap["normal"]


class Visitor:
    def defaultVisit(self, node):
        raise Exception(
            "INTERNAL ERROR: no visitor for node type `%s'" % (node.__class__.__name__)
        )

    def visitTranslationUnit(self, tu):
        for cxxInc in tu.cxxIncludes:
            cxxInc.accept(self)
        for inc in tu.includes:
            inc.accept(self)
        for su in tu.structsAndUnions:
            su.accept(self)
        for using in tu.builtinUsing:
            using.accept(self)
        for using in tu.using:
            using.accept(self)
        if tu.protocol:
            tu.protocol.accept(self)

    def visitCxxInclude(self, inc):
        pass

    def visitInclude(self, inc):
        # Note: we don't visit the child AST here, because that needs delicate
        # and pass-specific handling
        pass

    def visitStructDecl(self, struct):
        for f in struct.fields:
            f.accept(self)
        for a in struct.attributes.values():
            a.accept(self)

    def visitStructField(self, field):
        field.typespec.accept(self)

    def visitUnionDecl(self, union):
        for t in union.components:
            t.accept(self)
        for a in union.attributes.values():
            a.accept(self)

    def visitUsingStmt(self, using):
        for a in using.attributes.values():
            a.accept(self)

    def visitProtocol(self, p):
        for namespace in p.namespaces:
            namespace.accept(self)
        for mgr in p.managers:
            mgr.accept(self)
        for managed in p.managesStmts:
            managed.accept(self)
        for msgDecl in p.messageDecls:
            msgDecl.accept(self)
        for a in p.attributes.values():
            a.accept(self)

    def visitNamespace(self, ns):
        pass

    def visitManager(self, mgr):
        pass

    def visitManagesStmt(self, mgs):
        pass

    def visitMessageDecl(self, md):
        for inParam in md.inParams:
            inParam.accept(self)
        for outParam in md.outParams:
            outParam.accept(self)
        for a in md.attributes.values():
            a.accept(self)

    def visitParam(self, decl):
        for a in decl.attributes.values():
            a.accept(self)

    def visitTypeSpec(self, ts):
        pass

    def visitAttribute(self, a):
        if isinstance(a.value, Node):
            a.value.accept(self)

    def visitStringLiteral(self, sl):
        pass

    def visitDecl(self, d):
        for a in d.attributes.values():
            a.accept(self)


class Loc:
    def __init__(self, filename="<??>", lineno=0):
        assert filename
        self.filename = filename
        self.lineno = lineno

    def __repr__(self):
        return "%r:%r" % (self.filename, self.lineno)

    def __str__(self):
        return "%s:%s" % (self.filename, self.lineno)


Loc.NONE = Loc(filename="<??>", lineno=0)


class _struct:
    pass


class Node:
    def __init__(self, loc=Loc.NONE):
        self.loc = loc

    def accept(self, visitor):
        visit = getattr(visitor, "visit" + self.__class__.__name__, None)
        if visit is None:
            return getattr(visitor, "defaultVisit")(self)
        return visit(self)

    def addAttrs(self, attrsName):
        if not hasattr(self, attrsName):
            setattr(self, attrsName, _struct())


class NamespacedNode(Node):
    def __init__(self, loc=Loc.NONE, name=None):
        Node.__init__(self, loc)
        self.name = name
        self.namespaces = []

    def addOuterNamespace(self, namespace):
        self.namespaces.insert(0, namespace)

    def qname(self):
        return QualifiedId(self.loc, self.name, [ns.name for ns in self.namespaces])


class TranslationUnit(NamespacedNode):
    def __init__(self, type, name):
        NamespacedNode.__init__(self, name=name)
        self.filetype = type
        self.filename = None
        self.cxxIncludes = []
        self.includes = []
        self.builtinUsing = []
        self.using = []
        self.structsAndUnions = []
        self.protocol = None

    def addCxxInclude(self, cxxInclude):
        self.cxxIncludes.append(cxxInclude)

    def addInclude(self, inc):
        self.includes.append(inc)

    def addStructDecl(self, struct):
        self.structsAndUnions.append(struct)

    def addUnionDecl(self, union):
        self.structsAndUnions.append(union)

    def addUsingStmt(self, using):
        self.using.append(using)

    def setProtocol(self, protocol):
        self.protocol = protocol


class CxxInclude(Node):
    def __init__(self, loc, cxxFile):
        Node.__init__(self, loc)
        self.file = cxxFile


class Include(Node):
    def __init__(self, loc, type, name):
        Node.__init__(self, loc)
        suffix = "ipdl"
        if type == "header":
            suffix += "h"
        self.file = "%s.%s" % (name, suffix)


class UsingStmt(Node):
    def __init__(
        self,
        loc,
        cxxTypeSpec,
        cxxHeader=None,
        kind=None,
        attributes={},
    ):
        Node.__init__(self, loc)
        assert isinstance(cxxTypeSpec, QualifiedId)
        assert cxxHeader is None or isinstance(cxxHeader, str)
        assert kind is None or kind == "class" or kind == "struct"
        self.type = cxxTypeSpec
        self.header = cxxHeader
        self.kind = kind
        self.attributes = attributes

    def canBeForwardDeclared(self):
        return self.isClass() or self.isStruct()

    def isClass(self):
        return self.kind == "class"

    def isStruct(self):
        return self.kind == "struct"

    def isRefcounted(self):
        return "RefCounted" in self.attributes

    def isSendMoveOnly(self):
        moveonly = self.attributes.get("MoveOnly")
        return moveonly and moveonly.value in (None, "send")

    def isDataMoveOnly(self):
        moveonly = self.attributes.get("MoveOnly")
        return moveonly and moveonly.value in (None, "data")


# "singletons"


class PrettyPrinted:
    @classmethod
    def __hash__(cls):
        return hash_str(cls.pretty)

    @classmethod
    def __str__(cls):
        return cls.pretty


class ASYNC(PrettyPrinted):
    pretty = "async"


class INTR(PrettyPrinted):
    pretty = "intr"


class SYNC(PrettyPrinted):
    pretty = "sync"


class INOUT(PrettyPrinted):
    pretty = "inout"


class IN(PrettyPrinted):
    pretty = "in"


class OUT(PrettyPrinted):
    pretty = "out"


class Namespace(Node):
    def __init__(self, loc, namespace):
        Node.__init__(self, loc)
        self.name = namespace


class Protocol(NamespacedNode):
    def __init__(self, loc):
        NamespacedNode.__init__(self, loc)
        self.attributes = {}
        self.sendSemantics = ASYNC
        self.managers = []
        self.managesStmts = []
        self.messageDecls = []

    def nestedUpTo(self):
        if "NestedUpTo" not in self.attributes:
            return NOT_NESTED

        return NESTED_ATTR_MAP.get(self.attributes["NestedUpTo"].value, NOT_NESTED)

    def implAttribute(self, side):
        assert side in ("parent", "child")
        attr = self.attributes.get(side.capitalize() + "Impl")
        if attr is not None:
            return attr.value
        return None

    def procAttribute(self, side):
        assert side in ("parent", "child")
        attr = self.attributes.get(side.capitalize() + "Proc")
        if attr is not None:
            return attr.value
        elif side == "parent" and len(self.managers) == 0:
            return "Parent"  # Default for toplevel actors
        return None


class StructField(Node):
    def __init__(self, loc, type, name):
        Node.__init__(self, loc)
        self.typespec = type
        self.name = name


class StructDecl(NamespacedNode):
    def __init__(self, loc, name, fields, attributes):
        NamespacedNode.__init__(self, loc, name)
        self.fields = fields
        self.attributes = attributes
        # A list of indices into `fields` for determining the order in
        # which fields are laid out in memory.  We don't just reorder
        # `fields` itself so as to keep the ordering reasonably stable
        # for e.g. C++ constructors when new fields are added.
        self.packed_field_ordering = []


class UnionDecl(NamespacedNode):
    def __init__(self, loc, name, components, attributes):
        NamespacedNode.__init__(self, loc, name)
        self.components = components
        self.attributes = attributes


class Manager(Node):
    def __init__(self, loc, managerName):
        Node.__init__(self, loc)
        self.name = managerName


class ManagesStmt(Node):
    def __init__(self, loc, managedName):
        Node.__init__(self, loc)
        self.name = managedName


class MessageDecl(Node):
    def __init__(self, loc):
        Node.__init__(self, loc)
        self.name = None
        self.attributes = {}
        self.sendSemantics = ASYNC
        self.direction = None
        self.inParams = []
        self.outParams = []

    def addInParams(self, inParamsList):
        self.inParams += inParamsList

    def addOutParams(self, outParamsList):
        self.outParams += outParamsList

    def nested(self):
        if "Nested" not in self.attributes:
            return NOT_NESTED

        return NESTED_ATTR_MAP.get(self.attributes["Nested"].value, NOT_NESTED)

    def priority(self):
        if "Priority" in self.attributes:
            sourcePriority = self.attributes["Priority"].value
        else:
            sourcePriority = "normal"
        return priorityAttrMap.get(sourcePriority, NORMAL_PRIORITY)

    def replyPriority(self):
        if "ReplyPriority" in self.attributes:
            sourcePriority = self.attributes["ReplyPriority"].value
            if sourcePriority in priorityAttrMap:
                return priorityAttrMap[sourcePriority]
        return self.priority()


class Param(Node):
    def __init__(self, loc, typespec, name, attributes={}):
        Node.__init__(self, loc)
        self.name = name
        self.typespec = typespec
        self.attributes = attributes


class TypeSpec(Node):
    def __init__(self, loc, spec):
        Node.__init__(self, loc)
        assert isinstance(spec, str)
        self.spec = spec  # str
        self.array = False  # bool
        self.maybe = False  # bool
        self.nullable = False  # bool
        self.uniqueptr = False  # bool

    def basename(self):
        return self.spec

    def __str__(self):
        return self.spec


class Attribute(Node):
    def __init__(self, loc, name, value):
        Node.__init__(self, loc)
        self.name = name
        self.value = value


class StringLiteral(Node):
    def __init__(self, loc, value):
        Node.__init__(self, loc)
        self.value = value

    def __str__(self):
        return '"%s"' % self.value


class QualifiedId:  # FIXME inherit from node?
    def __init__(self, loc, baseid, quals=[]):
        assert isinstance(baseid, str)
        for qual in quals:
            assert isinstance(qual, str)

        self.loc = loc
        self.baseid = baseid
        self.quals = quals

    def qualify(self, id):
        self.quals.append(self.baseid)
        self.baseid = id

    def __str__(self):
        # NOTE: include a leading "::" in order to force all QualifiedIds to be
        # fully qualified types in C++
        return "::" + "::".join(self.quals + [self.baseid])


# added by type checking passes


class Decl(Node):
    def __init__(self, loc):
        Node.__init__(self, loc)
        self.progname = None  # what the programmer typed, if relevant
        self.shortname = None  # shortest way to refer to this decl
        self.fullname = None  # full way to refer to this decl
        self.loc = loc
        self.type = None
        self.scope = None
        self.attributes = {}

File: ipc/ipdl_new/ipdl/src/ast.rs
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

use std::fmt;
use std::collections::HashMap;

#[derive(Debug, Clone)]
pub struct QualifiedId {
    pub base_id: Identifier,
    pub quals: Vec<String>,
}

impl QualifiedId {
    pub fn new(base: Identifier) -> QualifiedId {
        QualifiedId { base_id: base, quals: Vec::new() }
    }

    pub fn qualify(mut self, id: Identifier) -> QualifiedId {
        self.quals.push(self.base_id.id);
        self.base_id = id;
        self
    }

    pub fn new_from_iter<'a, I> (mut ids: I) -> QualifiedId
        where I: Iterator<Item=&'a str>
    {
        let loc = Location { file_name: String::from("<builtin>"), lineno: 0, colno: 0 };
        let mut qual_id = QualifiedId::new(Identifier::new(String::from(ids.next().expect("Empty iterator when creating QID")), loc.clone()));
        for i in ids {
            qual_id = qual_id.qualify(Identifier::new(String::from(i), loc.clone()));
        }
        qual_id
    }

    pub fn short_name(&self) -> String {
        self.base_id.to_string()
    }

    pub fn full_name(&self) -> Option<String> {
        if self.quals.is_empty() {
            None
        } else {
            Some(self.to_string())
        }
    }

    pub fn loc(&self) -> &Location {
        &self.base_id.loc
    }
}

impl fmt::Display for QualifiedId {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        for q in &self.quals {
            try!(write!(f, "{}::", q));
        }
        write!(f, "{}", self.base_id)
    }
}

#[derive(Debug)]
pub struct TypeSpec {
    pub spec: QualifiedId,
    pub array: bool,
    pub nullable: bool,
}

impl TypeSpec {
    pub fn new(spec: QualifiedId) -> TypeSpec {
        TypeSpec { spec: spec, array: false, nullable: false }
    }

    pub fn loc(&self) -> &Location {
        self.spec.loc()
    }
}

#[derive(Debug)]
pub struct Param {
    pub name: Identifier,
    pub type_spec: TypeSpec,
}

#[derive(Debug)]
pub struct StructField {
    pub type_spec: TypeSpec,
    pub name: Identifier,
}

#[derive(Clone, Debug)]
pub struct Namespace {
    pub name: Identifier,
    pub namespaces: Vec<String>,
}

impl Namespace {
    pub fn qname(&self) -> QualifiedId {
        QualifiedId { base_id: self.name.clone(), quals: self.namespaces.clone() }
    }
}

#[derive(Debug, PartialEq, Clone, Copy)]
pub enum Compress {
    None,
    Enabled,
    All,
}

#[derive(Debug, PartialEq, Clone, Copy)]
pub enum SendSemantics {
    Async,
    Sync,
    Intr,
}

impl SendSemantics {
    pub fn is_async(&self) -> bool {
        self == &SendSemantics::Async
    }

    pub fn is_sync(&self) -> bool {
        self == &SendSemantics::Sync
    }

    pub fn is_intr(&self) -> bool {
        self == &SendSemantics::Intr
    }
}

#[derive(Debug, Clone, Copy, PartialEq, PartialOrd)]
pub enum Nesting {
    None,
    InsideSync,
    InsideCpow,
}

impl Nesting {
    pub fn is_none(&self) -> bool {
        self == &Nesting::None
    }

    pub fn inside_sync(&self) -> bool {
        self == &Nesting::InsideSync
    }

    pub fn inside_cpow(&self) -> bool {
        self == &Nesting::InsideCpow
    }
}

#[derive(Debug, Clone, Copy)]
pub enum Priority {
    Normal,
    High,
    Input,
}

#[derive(Debug, Clone, Copy, PartialEq)]
pub enum Direction {
    ToParent,
    ToChild,
    ToParentOrChild,
}

impl Direction {
    pub fn is_to_child(&self) -> bool {
        self == &Direction::ToChild
    }

    pub fn is_both(&self) -> bool {
        self == &Direction::ToParentOrChild
    }
}

#[derive(Debug, Clone)]
pub struct Location {
    pub file_name: String,
    pub lineno: usize,
    pub colno: usize,
}

impl fmt::Display for Location {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "{}:{}:{}", self.file_name, self.lineno, self.colno)
    }
}

#[derive(Debug, Clone)]
pub struct Identifier {
    pub id: String,
    pub loc: Location,
}

impl Identifier {
    pub fn new(name: String, loc: Location) -> Identifier {
        Identifier {
            id: name,
            loc: loc,
        }
    }
}

impl fmt::Display for Identifier {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "{}", self.id)
    }
}

#[derive(Debug)]
pub struct MessageDecl {
    pub name: Identifier,
    pub send_semantics: SendSemantics,
    pub nested: Nesting,
    pub prio: Priority,
    pub direction: Direction,
    pub in_params: Vec<Param>,
    pub out_params: Vec<Param>,
    pub compress: Compress,
    pub verify: bool,
}

#[derive(Debug)]
pub struct Protocol {
    pub send_semantics: SendSemantics,
    pub nested: Nesting,
    pub managers: Vec<Identifier>,
    pub manages: Vec<Identifier>,
    pub messages: Vec<MessageDecl>,
}

#[derive(Debug)]
pub enum CxxTypeKind {
  Struct,
  Class,
}

#[derive(Debug)]
pub struct UsingStmt {
    pub cxx_type: TypeSpec,
    pub header: String,
    pub kind: Option<CxxTypeKind>,
    pub refcounted: bool,
}

#[derive(Clone, Debug, PartialEq)]
pub enum FileType {
    Protocol,
    Header,
}

impl FileType {
    pub fn from_file_path(file_path: &str) -> Option<FileType> {
        if let Some(e) = file_path.rsplit('.').next() {
            if e == "ipdlh" {
                Some(FileType::Header)
            } else {
                Some(FileType::Protocol)
            }
        } else {
            None
        }
    }
}

// Translation unit identifier.
pub type TUId = i32;

#[derive(Debug)]
pub struct TranslationUnit {
    pub namespace: Namespace,
    pub file_type: FileType,
    pub file_name: String,
    pub cxx_includes: Vec<String>,
    pub includes: Vec<TUId>,
    pub using: Vec<UsingStmt>,
    pub structs: Vec<(Namespace, Vec<StructField>)>,
    pub unions: Vec<(Namespace, Vec<TypeSpec>)>,
    pub protocol: Option<(Namespace, Protocol)>,
}

pub struct AST {
    pub main_tuid: TUId,
    pub translation_units: HashMap<TUId, TranslationUnit>,
}
